File: src/cas/conductor.py
"""CAS Conductor — behavioral learning and user context.

Observes shell interactions over time and builds a persistent user profile
that is fed back into LLM system prompts, progressively calibrating CAS to
a specific user's patterns without requiring explicit configuration.

Profile is stored as JSON at ~/.cas/profile.json. Single-user design.
All observation methods are fail-safe — a corrupt or missing profile never
breaks the shell; the conductor simply returns no context.

Public API:
    conductor = Conductor()
    conductor.observe(intent, message, workspace=None, ws_type=None)
    conductor.user_context()  -> str for LLM system prompts
    conductor.profile_summary() -> dict for UI display
"""

from __future__ import annotations

import json
import logging
import re
from collections import Counter
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

logger = logging.getLogger(__name__)

_PROFILE_PATH = Path.home() / ".cas" / "profile.json"
_MAX_PHRASES      = 30
_MAX_CONTEXT_TYPES = 4
_MAX_CONTEXT_VERBS = 3

# Edit verbs — broader pattern, matches anywhere in message
_EDIT_VERB_RE = re.compile(
    r"\b(add|append|insert|revise|rewrite|remove|delete|update|change|modify|"
    r"expand|shorten|summarise|summarize|fix|improve|rename|refactor|"
    r"clean|polish|proofread|reorganize|restructure)\b",
    re.IGNORECASE,
)

# Document content-type nouns (for extracting topic from message text)
_DOC_TYPE_RE = re.compile(
    r"\b(resume|cv|letter|proposal|report|memo|essay|article|plan|outline|"
    r"note|brief|spec|story|blog|post|summary|agenda|budget|invoice|"
    r"contract|script|pitch|bio|profile|email|list|document|doc|"
    r"function|script|program|class|module|test|query|schema)\b",
    re.IGNORECASE,
)

# Minimum signal before generating any context
_MIN_WORKSPACES = 1   # lower threshold — one workspace is enough to start
_MIN_MESSAGES   = 2


class Conductor:
    """Observes CAS interactions and builds a persistent user profile."""

    def __init__(self, profile_path: Path = _PROFILE_PATH) -> None:
        self._path = profile_path
        self._profile: dict[str, Any] = self._load()

    # ── Persistence ─────────────────────────────────────────────

    def _load(self) -> dict[str, Any]:
        try:
            if self._path.exists():
                return self._defaults(json.loads(self._path.read_text()))
        except Exception as exc:
            logger.warning("Conductor: could not load profile: %s", exc)
        return self._defaults({})

    def _save(self) -> None:
        try:
            self._path.parent.mkdir(parents=True, exist_ok=True)
            self._path.write_text(json.dumps(self._profile, indent=2))
        except Exception as exc:
            logger.warning("Conductor: could not save profile: %s", exc)

    @staticmethod
    def _defaults(data: dict[str, Any]) -> dict[str, Any]:
        data.setdefault("doc_types", {})       # content-type nouns from messages
        data.setdefault("ws_types", {})        # actual workspace types (document/code/list)
        data.setdefault("edit_verbs", {})      # edit verb counts
        data.setdefault("phrases", [])         # recent create-intent messages
        data.setdefault("session_count", 0)
        data.setdefault("message_count", 0)
        data.setdefault("workspace_count", 0)  # actual workspaces created (not word count)
        data.setdefault("edit_count", 0)       # total edit operations
        data.setdefault("last_seen", None)
        return data

    # ── Observation ──────────────────────────────────────────────

    def observe(
        self,
        intent_kind: str,
        user_message: str,
        workspace_title: str | None = None,
        ws_type: str | None = None,
    ) -> None:
        """Record one shell interaction."""
        try:
            self._profile["message_count"] += 1
            self._profile["last_seen"] = datetime.now(timezone.utc).isoformat()

            if intent_kind == "create_workspace":
                self._observe_create(user_message, workspace_title, ws_type)
            elif intent_kind == "edit_workspace":
                self._observe_edit(user_message)

        except Exception as exc:
            logger.warning("Conductor.observe failed: %s", exc)
        finally:
            self._save()

    def observe_session_start(self) -> None:
        try:
            self._profile["session_count"] += 1
            self._save()
        except Exception as exc:
            logger.warning("Conductor.observe_session_start failed: %s", exc)

    def _observe_create(self, message: str, title: str | None, ws_type: str | None) -> None:
        # Increment actual workspace count once per creation
        self._profile["workspace_count"] += 1

        # Track actual workspace type (document/code/list)
        if ws_type:
            wt = self._profile["ws_types"]
            wt[ws_type] = wt.get(ws_type, 0) + 1

        # Extract topic nouns — deduplicated per message to avoid double-counting
        # "write a brief project plan" should count as one observation of 'plan',
        # not two (brief + plan both increment separately)
        types_found = list(dict.fromkeys(          # preserve order, deduplicate
            m.lower() for m in _DOC_TYPE_RE.findall(message)
        ))
        # Prefer the last/most-specific noun (usually the actual type)
        if types_found:
            primary = types_found[-1]
            doc_types = self._profile["doc_types"]
            doc_types[primary] = doc_types.get(primary, 0) + 1

        # Also record from title if it adds a different type
        if title:
            title_types = list(dict.fromkeys(m.lower() for m in _DOC_TYPE_RE.findall(title)))
            if title_types and title_types[-1] != (types_found[-1] if types_found else None):
                t = title_types[-1]
                doc_types = self._profile["doc_types"]
                doc_types[t] = doc_types.get(t, 0) + 1

        # Phrase ring buffer
        phrases = self._profile["phrases"]
        phrases.append(message.strip())
        if len(phrases) > _MAX_PHRASES:
            self._profile["phrases"] = phrases[-_MAX_PHRASES:]

    def _observe_edit(self, message: str) -> None:
        self._profile["edit_count"] = self._profile.get("edit_count", 0) + 1
        verbs_found = list(dict.fromkeys(m.lower() for m in _EDIT_VERB_RE.findall(message)))
        edit_verbs = self._profile["edit_verbs"]
        for v in verbs_found:
            edit_verbs[v] = edit_verbs.get(v, 0) + 1

    # ── Context generation ───────────────────────────────────────

    def user_context(self) -> str:
        """Return a natural-language user context string for LLM prompts.

        Returns empty string if insufficient signal.
        """
        try:
            return self._build_context()
        except Exception as exc:
            logger.warning("Conductor.user_context failed: %s", exc)
            return ""

    def _build_context(self) -> str:
        p = self._profile
        parts: list[str] = []

        # Lower threshold — start adapting after first workspace + a couple messages
        if p["workspace_count"] < _MIN_WORKSPACES and p["message_count"] < _MIN_MESSAGES:
            return ""

        # Workspace type preferences (document/code/list)
        ws_types = Counter(p.get("ws_types", {})).most_common(3)
        doc_types = Counter(p["doc_types"]).most_common(_MAX_CONTEXT_TYPES)

        if ws_types:
            dominant_ws_type = ws_types[0][0]
            if len(ws_types) == 1:
                parts.append(f"This user primarily creates {dominant_ws_type} workspaces.")
            else:
                type_list = ", ".join(f"{t} ({c})" for t, c in ws_types)
                parts.append(f"This user creates: {type_list} workspaces.")

        # Content topic preferences (proposal, resume, etc.)
        if doc_types:
            top = [t for t, _ in doc_types[:3]]
            if len(top) == 1:
                parts.append(f"Their documents are primarily {top[0]}s.")
            else:
                parts.append(
                    f"Their documents are typically {', '.join(top[:-1])} and {top[-1]}s."
                )

        # Edit style
        edit_verbs = Counter(p["edit_verbs"]).most_common(_MAX_CONTEXT_VERBS)
        edit_count = p.get("edit_count", 0)
        if edit_verbs and edit_count > 0:
            top_verbs = [v for v, _ in edit_verbs]
            if "rewrite" in top_verbs or "revise" in top_verbs:
                parts.append("They prefer full rewrites over incremental edits.")
            elif "shorten" in top_verbs or "summarize" in top_verbs or "summarise" in top_verbs:
                parts.append("They frequently ask to shorten or condense content.")
            elif "add" in top_verbs or "append" in top_verbs or "insert" in top_verbs:
                parts.append("They prefer adding sections over rewriting existing content.")
            elif "fix" in top_verbs or "improve" in top_verbs or "proofread" in top_verbs:
                parts.append("They frequently ask for quality improvements and fixes.")

        # Return user signal
        if p["session_count"] > 1:
            ws_str = f"{p['workspace_count']} workspace{'s' if p['workspace_count'] != 1 else ''}"
            parts.append(f"Returning user: {p['session_count']} sessions, {ws_str} created.")

        return " ".join(parts)

    # ── Introspection ─────────────────────────────────────────────

    def profile_summary(self) -> dict[str, Any]:
        """Return profile plus the current generated context string."""
        summary = dict(self._profile)
        summary["context"] = self.user_context()
        summary["has_context"] = bool(summary["context"])
        return summary

    def reset(self) -> None:
        self._profile = self._defaults({})
        self._save()

File: internal/conductor/conductor.go
// Package conductor implements behavioral learning for CAS.
//
// The Conductor observes shell interactions over time and builds a persistent
// user profile stored at ~/.cas/profile.json. On each interaction it updates
// the profile and derives a natural-language context string that is appended
// to LLM system prompts, progressively calibrating CAS to the specific user
// without requiring explicit configuration.
//
// All methods are fail-safe: a corrupt or missing profile never breaks the
// shell; observe errors are logged and silently discarded.
package conductor

import (
	"encoding/json"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"
)

// Thresholds before context generation starts.
const (
	minWorkspaces = 1
	minMessages   = 2
	maxPhrases    = 30
	maxCtxTypes   = 4
	maxCtxVerbs   = 3
)

// Profile is the persisted behavioral learning state.
type Profile struct {
	DocTypes       map[string]int `json:"doc_types"`       // content-type nouns
	WSTypes        map[string]int `json:"ws_types"`        // document/code/list counts
	EditVerbs      map[string]int `json:"edit_verbs"`      // edit verb counts
	Phrases        []string       `json:"phrases"`         // recent create-intent messages
	SessionCount   int            `json:"session_count"`
	MessageCount   int            `json:"message_count"`
	WorkspaceCount int            `json:"workspace_count"` // workspaces created
	EditCount      int            `json:"edit_count"`
	LastSeen       string         `json:"last_seen"` // RFC3339
}

func defaultProfile() Profile {
	return Profile{
		DocTypes:  make(map[string]int),
		WSTypes:   make(map[string]int),
		EditVerbs: make(map[string]int),
		Phrases:   []string{},
	}
}

// Conductor observes CAS interactions and builds a persistent user profile.
type Conductor struct {
	path    string
	profile Profile
}

// New returns a Conductor backed by the given profile path.
// Pass an empty string to use the default ~/.cas/profile.json.
func New(profilePath string) *Conductor {
	if profilePath == "" {
		home, _ := os.UserHomeDir()
		profilePath = filepath.Join(home, ".cas", "profile.json")
	}
	c := &Conductor{path: profilePath}
	c.profile = c.load()
	return c
}

// ── Persistence ───────────────────────────────────────────────────

func (c *Conductor) load() Profile {
	data, err := os.ReadFile(c.path)
	if err != nil {
		return defaultProfile()
	}
	var p Profile
	if err := json.Unmarshal(data, &p); err != nil {
		log.Printf("conductor: corrupt profile, resetting: %v", err)
		return defaultProfile()
	}
	// Ensure maps are non-nil (old profiles may predate a field)
	if p.DocTypes == nil {
		p.DocTypes = make(map[string]int)
	}
	if p.WSTypes == nil {
		p.WSTypes = make(map[string]int)
	}
	if p.EditVerbs == nil {
		p.EditVerbs = make(map[string]int)
	}
	if p.Phrases == nil {
		p.Phrases = []string{}
	}
	return p
}

func (c *Conductor) save() {
	if err := os.MkdirAll(filepath.Dir(c.path), 0700); err != nil {
		log.Printf("conductor: mkdir: %v", err)
		return
	}
	data, err := json.MarshalIndent(c.profile, "", "  ")
	if err != nil {
		log.Printf("conductor: marshal: %v", err)
		return
	}
	if err := os.WriteFile(c.path, data, 0600); err != nil {
		log.Printf("conductor: write: %v", err)
	}
}

// ── Patterns ──────────────────────────────────────────────────────

var editVerbRe = regexp.MustCompile(
	`(?i)\b(add|append|insert|revise|rewrite|remove|delete|update|change|modify|` +
		`expand|shorten|summarise|summarize|fix|improve|rename|refactor|` +
		`clean|polish|proofread|reorganize|restructure)\b`,
)

var docTypeRe = regexp.MustCompile(
	`(?i)\b(resume|cv|letter|proposal|report|memo|essay|article|plan|outline|` +
		`note|brief|spec|story|blog|post|summary|agenda|budget|invoice|` +
		`contract|script|pitch|bio|profile|email|list|document|doc|` +
		`function|program|class|module|test|query|schema)\b`,
)

// ── Observation ───────────────────────────────────────────────────

// Observe records one shell interaction.
// intentKind is one of "create_workspace", "edit_workspace", "close_workspace", "chat".
func (c *Conductor) Observe(intentKind, message, workspaceTitle, wsType string) {
	defer func() {
		if r := recover(); r != nil {
			log.Printf("conductor: observe panic: %v", r)
		}
		c.save()
	}()

	c.profile.MessageCount++
	c.profile.LastSeen = time.Now().UTC().Format(time.RFC3339)

	switch intentKind {
	case "create_workspace":
		c.observeCreate(message, workspaceTitle, wsType)
	case "edit_workspace":
		c.observeEdit(message)
	}
}

// ObserveSessionStart increments the session counter.
func (c *Conductor) ObserveSessionStart() {
	defer c.save()
	c.profile.SessionCount++
}

func (c *Conductor) observeCreate(message, title, wsType string) {
	c.profile.WorkspaceCount++

	if wsType != "" {
		c.profile.WSTypes[wsType]++
	}

	// Extract topic nouns — deduplicated per message, prefer last (most specific)
	found := dedupeLower(docTypeRe.FindAllString(message, -1))
	if len(found) > 0 {
		primary := found[len(found)-1]
		c.profile.DocTypes[primary]++
	}

	// Also observe from title if it adds a different noun
	if title != "" {
		titleFound := dedupeLower(docTypeRe.FindAllString(title, -1))
		if len(titleFound) > 0 {
			t := titleFound[len(titleFound)-1]
			if len(found) == 0 || t != found[len(found)-1] {
				c.profile.DocTypes[t]++
			}
		}
	}

	// Phrase ring buffer
	c.profile.Phrases = append(c.profile.Phrases, strings.TrimSpace(message))
	if len(c.profile.Phrases) > maxPhrases {
		c.profile.Phrases = c.profile.Phrases[len(c.profile.Phrases)-maxPhrases:]
	}
}

func (c *Conductor) observeEdit(message string) {
	c.profile.EditCount++
	verbs := dedupeLower(editVerbRe.FindAllString(message, -1))
	for _, v := range verbs {
		c.profile.EditVerbs[v]++
	}
}

// ── Context generation ────────────────────────────────────────────

// UserContext returns a natural-language string for LLM system prompts.
// Returns empty string when there is insufficient signal.
func (c *Conductor) UserContext() string {
	defer func() {
		if r := recover(); r != nil {
			log.Printf("conductor: user_context panic: %v", r)
		}
	}()
	return c.buildContext()
}

func (c *Conductor) buildContext() string {
	p := c.profile
	if p.WorkspaceCount < minWorkspaces && p.MessageCount < minMessages {
		return ""
	}

	var parts []string

	// Workspace type preferences
	wsTypes := topN(p.WSTypes, 3)
	docTypes := topN(p.DocTypes, maxCtxTypes)

	if len(wsTypes) > 0 {
		dominant := wsTypes[0].key
		if len(wsTypes) == 1 {
			parts = append(parts, "This user primarily creates "+dominant+" workspaces.")
		} else {
			var items []string
			for _, kv := range wsTypes {
				items = append(items, kv.key)
			}
			parts = append(parts, "This user creates: "+strings.Join(items, ", ")+" workspaces.")
		}
	}

	// Content topic preferences
	if len(docTypes) > 0 {
		tops := make([]string, len(docTypes))
		for i, kv := range docTypes {
			tops[i] = kv.key
		}
		if len(tops) == 1 {
			parts = append(parts, "Their documents are primarily "+tops[0]+"s.")
		} else {
			last := tops[len(tops)-1]
			rest := tops[:len(tops)-1]
			parts = append(parts, "Their documents are typically "+strings.Join(rest, ", ")+" and "+last+"s.")
		}
	}

	// Edit style
	editVerbs := topN(p.EditVerbs, maxCtxVerbs)
	if len(editVerbs) > 0 && p.EditCount > 0 {
		verbSet := make(map[string]bool)
		for _, kv := range editVerbs {
			verbSet[kv.key] = true
		}
		switch {
		case verbSet["rewrite"] || verbSet["revise"]:
			parts = append(parts, "They prefer full rewrites over incremental edits.")
		case verbSet["shorten"] || verbSet["summarize"] || verbSet["summarise"]:
			parts = append(parts, "They frequently ask to shorten or condense content.")
		case verbSet["add"] || verbSet["append"] || verbSet["insert"]:
			parts = append(parts, "They prefer adding sections over rewriting existing content.")
		case verbSet["fix"] || verbSet["improve"] || verbSet["proofread"]:
			parts = append(parts, "They frequently ask for quality improvements and fixes.")
		}
	}

	// Return user signal
	if p.SessionCount > 1 {
		ws := "workspace"
		if p.WorkspaceCount != 1 {
			ws = "workspaces"
		}
		parts = append(parts, "Returning user: "+itoa(p.SessionCount)+" sessions, "+itoa(p.WorkspaceCount)+" "+ws+" created.")
	}

	return strings.Join(parts, " ")
}

// ── Introspection ─────────────────────────────────────────────────

// ProfileSummary returns the profile plus the current context string.
func (c *Conductor) ProfileSummary() map[string]interface{} {
	ctx := c.UserContext()
	return map[string]interface{}{
		"doc_types":       c.profile.DocTypes,
		"ws_types":        c.profile.WSTypes,
		"edit_verbs":      c.profile.EditVerbs,
		"session_count":   c.profile.SessionCount,
		"message_count":   c.profile.MessageCount,
		"workspace_count": c.profile.WorkspaceCount,
		"edit_count":      c.profile.EditCount,
		"last_seen":       c.profile.LastSeen,
		"context":         ctx,
		"has_context":     ctx != "",
	}
}

// Reset wipes the profile and saves.
func (c *Conductor) Reset() {
	c.profile = defaultProfile()
	c.save()
}

// ── Helpers ───────────────────────────────────────────────────────

type kv struct {
	key   string
	count int
}

// topN returns the top n entries from a frequency map, sorted by count desc.
func topN(m map[string]int, n int) []kv {
	out := make([]kv, 0, len(m))
	for k, v := range m {
		out = append(out, kv{k, v})
	}
	sort.Slice(out, func(i, j int) bool {
		if out[i].count != out[j].count {
			return out[i].count > out[j].count
		}
		return out[i].key < out[j].key // stable alphabetical tiebreak
	})
	if len(out) > n {
		out = out[:n]
	}
	return out
}

// dedupeLower lowercases and deduplicates a slice preserving order.
func dedupeLower(words []string) []string {
	seen := make(map[string]bool)
	var out []string
	for _, w := range words {
		l := strings.ToLower(w)
		if !seen[l] {
			seen[l] = true
			out = append(out, l)
		}
	}
	return out
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var buf [20]byte
	pos := len(buf)
	for n > 0 {
		pos--
		buf[pos] = byte(n%10) + '0'
		n /= 10
	}
	if neg {
		pos--
		buf[pos] = '-'
	}
	return string(buf[pos:])
}
